### PR TITLE
`neutral_mixed` MMS tests and expanded differential operator MMS tests

### PR DESCRIPTION
**Pull request by mrhardman**
The changes included here aim to significantly expand the tests available in `tests/mms`. The following features are added:

 - symbolic operators in the `tests/mms/perpendicular_laplacian.py` module covering all of the differential operators that are present in `neutral_mixed`.
 - extension of `nonorthogonal_test.py` and `orthogonal_test.py` to test the newly added differential operators. This requires some changes to `main.cxx`.
 - an automatic test of the right hand side of the equation solved in `neutral_mixed`. We check that the definition of the right hand side matches the expected convergence order. Note that the expected convergence order is second order (2) when `evolve_momentum = false`, but the observed convergence order is first order or worse when `evolve_momentum = true`.
 - ability to add an external source of momentum to neutrals in `neutral_mixed`.
 - ability to read normalised sources in `neutral_mixed`.

The major addition is the test script `neutral_mixed_ddt_test.py`, which runs the following four tests with prescribed functions `Nd`, `Pd`, `NVd` that are constructed to be functions of `(x,y)` and satisfy the boundary conditions imposed in `neutral_mixed`.

 - `neutral_mixed/evolve_momentum_False_conservation_test_False`. Check the definition of the RHS when `evolve_momentum = false`, by calculating symbolically `ddt(Nd)` and `ddt(Pd)`, and using these symbolic expressions as sources to cancel the numerical `ddt(Nd)` and `ddt(Pd)`, such that the output `ddt(Nd)` and `ddt(Pd)` are of order the discretisation error. The spatial resolution is scanned to check convergence, with the following results.
fig_0
fig_1

-  `neutral_mixed/evolve_momentum_False_conservation_test_True`. Check that the RHS when `evolve_momentum = false` conserves `Nd` and `Pd`. The spatial resolution is scanned to check convergence, with the following results. Note that density is exactly conserved but pressure is only conserved within numerical error. A volume integration is done over `ddt(Nd)` and `ddt(Pd)` (without a symbolic source) to obtain these errors.
fig_0
fig_1

 - `neutral_mixed/evolve_momentum_True_conservation_test_False`. Check the definition of the RHS when `evolve_momentum =true`, by calculating symbolically `ddt(Nd)`, `ddt(NVd)` and `ddt(Pd)`, and using these symbolic expressions as sources to cancel the numerical `ddt(Nd)`, `ddt(NVd)`and `ddt(Pd)`, such that the output `ddt(Nd)`, `ddt(NVd)`and `ddt(Pd)` are of order the discretisation error. The spatial resolution is scanned to check convergence, with the following results.
fig_1
fig_2
fig_0

- `neutral_mixed/evolve_momentum_True_conservation_test_True`. Check that the RHS when `evolve_momentum = true` conserves `Nd` and energy. The spatial resolution is scanned to check convergence, with the following results. Note that conservation is first order accurate for density and less than first order for energy.
fig_0
fig_1

These tests can be readily extended to test the `ddt()` of other physics modules, or to 3D `(x,y,z)`, allowing for rapid testing of the definition of the equations solved.

Potential issues requiring further work:
 - [ ] Flag documentation issues, with respect to dimensionally incorrect documentation for momentum equation for neutrals, factors of mass in viscous coefficients.
 - [ ] Determine why energy conservation was not observed.
 - [ ] Extend nonorthogonal operators to charged species.
 - [ ] Demonstrate `neutral_mixed` working smoothly with nonorthogonal mesh in realistic physics case.
 - [ ] Improve readme and documentation for tests  -- reviewers please make suggestions.
 - [ ] Anything else raised by reviewers.

### CODE REVIEW

**Comment by cmacmackin** [comment]
The reason the CIs keep failing on this is that you're using the CVODE solver but SUNDIALS isn't available in your CI image. I've switched it over to PVODE, which seems to work. I've also updated the MMS testing script to print the output when there is an error, which should make such issues easier to diagnose in future.

**Comment by mikekryjak** [comment]
Thanks @cmacmackin for your investigation so far. Maybe @ZedThree could chime in once he's back from leave - this seems particularly subtle..

**Review by ZedThree — CHANGES_REQUESTED** [review]

File: src/neutral_mixed.cxx [hunk 1]
```diff
@@ -283,14 +311,17 @@ void NeutralMixed::finally(const Options& state) {
...
-    // Dnn = Vth^2 / sigma
-    Dnn = (Tn / AA) / (get<Field3D>(localstate["collision_frequency"]) + Rnn);
+  if (rnn_override > 0.0) { 
+    // user has set an override for collision frequency
+    Dnn = (Tn / AA) / rnn_override;
   } else {
-    Dnn = (Tn / AA) / Rnn;
+    if (localstate.isSet("collision_frequency")) {
+      // Dnn = Vth^2 / sigma
+      Dnn = (Tn / AA) / (get<Field3D>(localstate["collision_frequency"]) + Rnn);
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/neutral_mixed.cxx:320]: Should the new user-override also add on to `collision_frequency`?

[the remark this answers is not in the document]
Comment (mikekryjak) [reply src/neutral_mixed.cxx:320]: The user input is actually `neutral_lmax`, the max mean free path length scale. `Rnn` does not need to be a user input.
Both variables are poorly named, by the way. I have reworked this on my fork which improves the flux limitation for the neutral model which is due testing and merging once I get the time.

File: src/neutral_mixed.cxx [hunk 2]
```diff
@@ -126,7 +144,17 @@ NeutralMixed::NeutralMixed(const std::string& name, Options& alloptions, Solver*
...
                         .withDefault(pressure_source)
-                    / (SI::qe * Nnorm * Tnorm * Omega_ci);
+                    / pressure_norm;
+  // Try to read the momentum source from the mesh
+  momentum_source = 0.0;
+  mesh->get(momentum_source, std::string("NV") + name + "_src");
+   // Allow the user to override the source
+  momentum_source = alloptions[std::string("NV") + name]["source"]
+                        .doc(std::string("Source term in ddt(NV") + name
+                             + std::string("). Units [kg m^-2 s^-2]"))
```
[the remark this answers is not in the document]
Comment (mrhardman) [reply src/neutral_mixed.cxx]: Resolved by 1e0d13b542fae82ff09bf3d422739536a750224f

File: tests/mms/main.cxx [hunk 3]
```diff
@@ -25,13 +52,45 @@ class nameandfunction4 {
...
+const auto differential_operators_1_arg = {
+    nameandfunction1{"Div_par(f)",
+                 [](const Field3D &f) {
+                   return Div_par(f);
+                 }},
+    nameandfunction1{"Grad_par(f)",
+                 [](const Field3D &f) {
+                   return Grad_par(f);
+                 }},
+};
```
[the remark this answers is not in the document]
Comment (mrhardman) [reply tests/mms/main.cxx:65]: Attempting to do this leads to error messages like 
```
error: cannot resolve overloaded function 'Div_par' based on conversion to type 'std::function<Field3D(const Field3D&)>'
   57 |     nameandfunction1{"Div_par(f)", Div_par},
```
In principle, I happy for this code to be simplified if possible, if you wanted to make a PR with working changes.
